textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -124,10 +140,62 @@ public ITuple getNextTuple() throws DataFlowException {
                     }
 
                     boolean allTokenPresent = areAllTrue(tokensPresent);
+                    if(predicate.getOperatorType() == KeywordOperatorType.BASIC) {
+
+                        if (!allTokenPresent) {
+                            spanList.removeAll(spanForThisField);
+                        }
+                    }
+
+                    else if(predicate.getOperatorType() == KeywordOperatorType.PHRASE){
 
-                    if (!allTokenPresent) {
                         spanList.removeAll(spanForThisField);
+                        //relevantWordsInQuery has all the tokens (including duplicates but excluding stop words) in the query.
+                        List<String> relevantWordsInQuery = new ArrayList<>();
+                        //relevantWordsInQueryOffset maintains the offset position of tokens in query
+                        List<Integer> relevantWordsInQueryOffset = new ArrayList<>();
+                        String[] queryArray = query.split(" ");
+                        if (allTokenPresent) {
+                            for(int iter = 0; iter < queryArray.length; iter++){
+                                if(queryTokens.contains(queryArray[iter])){
```

Comment:
queryTokens are split using analyzer. queryArray needs to be to split with spaces.

if the query is "lin and lin and angry". queryTokens will have only "lin" and "angry". queryArray will have "lin","and","lin","and","angry". from this queryArray and queryToken we obtain the semantics for phrase query in following loop.

for(int iter = 0; iter < queryArray.length; iter++){
                                if(queryTokens.contains(queryArray[iter])){
                                    relevantWordsInQuery.add(queryArray[iter]);
                                    relevantWordsInQueryOffset.add(iter);
                                }
                            }